src/potentials.h

```diff
@@ -8,27 +8,138 @@
+//! type of a function defining a combination rule of a heterogeneous pair interaction
+typedef std::function<double(double, double)> TCombinatorFunc;
+//! type of a function modifying combinator's output
+typedef std::function<double(double)> TModifierFunc;
+
+/**
+ * @brief Data for a custom (heterogeneous) interaction between two given atom types.
+ *
+ * The very same format is used as for a homogeneous interaction specified directly on an atom type.
+ */
+struct InteractionData {
+    std::array<AtomData::Tid, 2> atom_id;
+    AtomData interaction;
+};
+
+void from_json(const json &j, std::vector<InteractionData> &interactions);
+void to_json(json &j, const std::vector<InteractionData> &interactions);
+
+/**
+ * @brief Known named combination rules for parameters of pair potential interaction.
+ *
+ * When adding a new one, add a json mapping. Also consider appending the PairMixer::getCombinator()
+ * method to recognize the new rule.
+ */
+enum ECombinationRule {COMB_UNDEFINED = 0, COMB_ARITHMETIC, COMB_GEOMETRIC, COMB_LORENTZ_BERTHELOT};
+NLOHMANN_JSON_SERIALIZE_ENUM( ECombinationRule, {
+    {COMB_UNDEFINED, "undefined"},
+    {COMB_ARITHMETIC, "arithmetic"},
+    {COMB_GEOMETRIC, "geometric"},
+    {COMB_LORENTZ_BERTHELOT, "lorentz_berthelot"},
+    {COMB_LORENTZ_BERTHELOT, "LB"}, // alternative non-canonical name
+})
+
+/**
+ * @brief Exception for handling pair potential initialization.
+ */
+struct PairPotentialException : public std::runtime_error {
+    PairPotentialException(const std::string msg) : std::runtime_error (msg) {};
+};
+
+/**
+ * @brief PairMixer creates a matrix of pair potential coefficients based on the atom properties
+ * and/or custom values using an arbitrary combination rule.
+ *
+ * PairMixer holds three functions that are applied in order extractor → combinator → modifier to create
+ * a coefficient matrix for all possible interactions. The function createPairMatrix applies the functions
+ * on all atom type pairs, and optionally also on the list of custom pair parameters (not the combinator
+ * function).
+ */
+class PairMixer {
+    TExtractorFunc extractor;   //!< Function extracting the coefficient from the AtomData structure
+    TCombinatorFunc combinator; //!< Function combining two values
+    TModifierFunc modifier;     //!< Function modifying the result for fast computations, e.g., a square of
+
+  public:
+    PairMixer(TExtractorFunc extractor, TCombinatorFunc combinator, TModifierFunc modifier = &modIdentity) :
+            extractor(extractor), combinator(combinator), modifier(modifier) {};
+
+    //! @return a square matrix of atoms.size()
+    TPairMatrixPtr createPairMatrix(const std::vector<AtomData> &atoms);
+    //! @return a square matrix of atoms.size()
+    TPairMatrixPtr createPairMatrix(const std::vector<AtomData> &atoms,
+                                    const std::vector<InteractionData> &interactions);
+
+    enum ECoefficient {COEF_ANY, COEF_SIGMA, COEF_EPSILON};
```

Comment:
Skip the `E` prefix or spell out as `EnumCoefficient`

Reply:
Renamed to `PotentialCoefficient`.